### DISCUSSION TITLE
What is the purpose of Demo project?

### DISCUSSION THREAD

**Discussion by jennyhougen** [FHIDev/Fhi.HelseId.Demo#17] [General]
The demo solution have been restructured and there is currently one sample with a Angular client and belonging backend calling an api. It is in .Net 8. But we are targeting .Net 8 and .Net 9 so we need to be able to test both versions. 

We need to discuss the following:
- Should the demo application only be at the latests versions or should we also demonstrate supported versions?  
- Should we have another solution in HelseId repo where we test different setup and versions and only have the latest version targeted in the demo solution?

**Comment by OsirisTerje** [comment]
The purpose of the Demo project(s) is to show **how** to use the components in different scenarios.  The users of the demo project are for other developers.
Thus they have to be simple, and whatever code we add to support the packages should be commented. 
The demo project should have **that** focus, and not a verification focus.

Imho, I believe it is enough to just have support for the latest , but in the comments/readme/wiki state which other versions are supported.  
If there are special things that are only supported in a specific version, that should of course be pointed out,  like PAR, which we only support in .net 9.  We should then also state when that is important  or not.

The 2nd point about other versions - we could do that, but I am not sure it should be in the demo repo. It may be, but the risk of overcomplicating things increases.  If we just want to be sure ourselves, that is for verification or testing purposes, it should perhaps be placed in a separate repo.

**Comment by martinmine** [comment]
Which version of .NET the user of our packages is using is an implementation detail, it should not be a big difference in terms of how they use Fhi.HelseId if they use .NET 8 or .NET 9. Currently it is only PAR which they might care about which they can use in .NET 9. Ideally when teams maintain their services, they should be quite up to date with their NuGet dependencies, and when they upgrade from .NET 8 to .NET 9 the only change they would do is the `TargetFramework` attribute in their .csproj file(s). The developers should not have to think about Fhi.HelseId, or any other dependencies for that matter as this should be the only change they would have to do to move to a newer .NET version. While there may be other packages available to them after upgrading to .NET 9, those are new options that are available after upgrading .NET. If they had to take a lot into account when updating between .NET versions, it would make more sense for us to have this explicit in our demo projects (think older EntityFramework versions back in the ".NET Framework days").
Since the approach in modern .NET is to have smaller version changes between each .NET version, and a new major .NET release does not lead to a lot of significant changes, it makes more sense for us to treat this as an implementation detail. Therefore I think we should have our demo projects up to date, and be explicit if we have a feature that is bound to a specific .NET version (such as PAR - which we have been). I would picture a demo project as sort of a reference project where teams would go to answer the question "How can I use Fhi.HelseId in my use-case". Which version of Fhi.HelseId they will use then is not that important as we expect them to use the latest. The same goes for .NET as long as it is a version that is still supported.

> Should we have another solution in HelseId repo where we test different setup and versions and only have the latest version targeted in the demo solution?

Is this something we can achieve in the tests we have in Fhi.HelseId? I also would like to know what we want to achieve with version-dependent tests. If we actually need tests for specific versions of .NET we should take a step back and look at if there are other aspects we could adjust to avoid having those kinds of tests - both due to the implementation detail I mentioned above and that it is a signal we are making things more complicated than they have to.

**Comment by jennyhougen** [comment]
You have some good points. I agree we should let the Demo project be for the developers. I will remove the .Net spesific target version in the name. But we need a place where we can test the different targets. As you say @martinmine maybe this can be done in an integration test. Need to explore that. I will create a task.
